storage/src/main/java/org/apache/uniffle/storage/handler/impl/HdfsClientReadHandler.java

```diff
@@ -127,21 +127,32 @@ public ShuffleDataResult readShuffleData() {
       init(fullShufflePath);
     }
 
-    if (readHandlerIndex >= readHandlers.size()) {
+    if (finished()) {
       return new ShuffleDataResult();
     }
 
-    HdfsShuffleReadHandler hdfsShuffleFileReader = readHandlers.get(readHandlerIndex);
-    ShuffleDataResult shuffleDataResult = hdfsShuffleFileReader.readShuffleData();
-
-    while (shuffleDataResult == null) {
-      ++readHandlerIndex;
+    HdfsShuffleReadHandler hdfsShuffleFileReader;
+    ShuffleDataResult shuffleDataResult = null;
+    do {
       if (readHandlerIndex >= readHandlers.size()) {
         return new ShuffleDataResult();
       }
       hdfsShuffleFileReader = readHandlers.get(readHandlerIndex);
-      shuffleDataResult = hdfsShuffleFileReader.readShuffleData();
-    }
+      if (hdfsShuffleFileReader.finished()) {
+        ++readHandlerIndex;
+        continue;
+      }
+      try {
+        shuffleDataResult = hdfsShuffleFileReader.readShuffleData();
+      } catch (Exception e) {
+        ++readHandlerIndex;
+        continue;
+      }
+
+      if (shuffleDataResult == null || shuffleDataResult.isEmpty()) {
```

Comment:
It looks strange to use the same condition `(shuffleDataResult == null || shuffleDataResult.isEmpty())` twice.